### PR TITLE
Add live preview to guidance markdown editor

### PR DESCRIPTION
**Pull request by DavidBiddle**
### What problem does this pull request solve?

Trello card: https://trello.com/c/cUoS52eU/986-enhance-markdown-editing-with-live-javascript-preview

Adds:
- a new endpoint on the guidance controller that takes markdown and returns the rendered HTML in its JSON response
- a new JS module which sends the contents of a textarea to this endpoint and renders the content in a preview area
  - the JS will submit to the endpoint when:
    - the page first loads
    - when the user types (we debounce it to 1 second - i.e. the request won't fire until the user has stopped typing for a second, so we don't accidentally DDoS ourselves by sending a request for every single keystroke)
    - when the user presses one of the toolbar buttons

### Things to consider when reviewing

- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated?

### CODE REVIEW

File: app/controllers/pages/additional_guidance_controller.rb [hunk 1]
```diff
@@ -47,6 +47,13 @@ def update
     end
   end
 
+  def render_preview
+    markdown = params[:guidance_markdown]
+    preview_html = markdown.present? ? GovukFormsMarkdown.render(markdown) : "<p>You have no content to preview</p>"
```
Comment (aliuk2012) [inline app/controllers/pages/additional_guidance_controller.rb:52]: Could you not maybe create a new instance the guidance_form which just sets `guidance_markdown` and then reuse `preview_html` method?

File: spec/requests/pages/additional_guidance_controller_spec.rb [hunk 2]
```diff
@@ -195,4 +195,20 @@
...
   end
+
+  describe "#render_preview" do
+    let(:guidance_markdown) { "### Markdown" }
+
+    before do
+      post additional_guidance_render_preview_path(form_id: form.id), params: { guidance_markdown: }
+    end
+
+    it "returns a JSDON object containing the converted HTML" do
```
Comment (aliuk2012) [inline spec/requests/pages/additional_guidance_controller_spec.rb:206]:
```suggestion
    it "returns a JSON object containing the converted HTML" do
```

File: app/controllers/pages/additional_guidance_controller.rb (hunk 1 above)
Comment (DavidBiddle) [reply app/controllers/pages/additional_guidance_controller.rb]: @aliuk2012 Good suggestion, have updated!

File: app/frontend/javascript/ajax-markdown-preview/index.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  retryButton.classList.add('govuk-button', 'govuk-button--secondary')
+  retryButton.innerHTML = 'Retry preview'
+  addEventListeners(retryButton, manuallyTriggerMarkdownPreview)
+  store.target.appendChild(retryButton)
+}
+
+const triggerAjaxMarkdownPreview = async () => {
+  try {
+    const response = await window.fetch(
+      '/forms/16/pages/new/additional-guidance-preview',
```
Comment (aliuk2012) [inline app/frontend/javascript/ajax-markdown-preview/index.js:30]: should this be dynamic?

Comment (DavidBiddle) [reply app/frontend/javascript/ajax-markdown-preview/index.js:30]: Yes! Good spot 😅

Comment (DavidBiddle) [reply app/frontend/javascript/ajax-markdown-preview/index.js]: Should be fixed now

**Review by aliuk2012 — APPROVED** [review]
Nicely done 👍

**Pull request merged**